openapi3/visited.go

```diff
@@ -0,0 +1,81 @@
+package openapi3
+
+import "context"
+
+func newVisited() *visitedComponent {
+	return &visitedComponent{
+		header: make(map[*Header]struct{}),
+		schema: make(map[*Schema]struct{}),
+	}
+}
+
+type visitedComponent struct {
+	header map[*Header]struct{}
+	schema map[*Schema]struct{}
+}
+
+type ctxKey struct{}
+
+// withContext returns a copy of ctx with visitedComponent associated. If an instance
+// of visitedComponent is already in the context, the context is not updated.
+//
+// Returned ctx can be passed to children function calls and used to check whether
+// component already visited. For instance:
+//
+//	ctx := context.Background()
+//	ctx = newVisited().withContext(ctx)
+//	...
+//	doc.deferSchemaRecursively(ctx, schema)
+//	func (doc *T) deferSchemaRecursively(ctx context.Context, s *Schema) {
+//	    if s == nil || isVisitedSchema(ctx, s) {
+//	        return
+//	    }
+//	}
+func (v *visitedComponent) withContext(ctx context.Context) context.Context {
```

Comment:
I'm not sure we should be adding things to ctx. What made you store these maps in here? Could this be moved to the root document type T instead?

Thanks

Reply:
It can be moved to document type T, but it have one minus: changing document we change it globally. It means, that it's not thread safe. Context allows us to restrict recursion only for this exact call of `InternalizeRef`.

If you ok with that minus, I can add the maps to root document type T and clean them before exiting `InternalizeRef`.
What do you decide?

Thanks

Reply:
Moved the visitedComponent to doc *T.
Could you please check?